### PR TITLE
[ADD] product_contract: Create module

### PR DESCRIPTION
**Pull request by tedsalmon**
* Add contract functionality to `product.templates`
* Add logic to create contracts from `sale.order` that contains contract products.

Depends: 

~~- [x] #42~~

cc @lasley

### CODE REVIEW

File: product_contract/models/product_template.py
Comment (pedrobaeza) [inline product_contract/models/product_template.py]: Set it to `False`, instead of `None`, that is the standard "null" value in Odoo.

File: product_contract/models/sale_order.py
Comment (pedrobaeza) [inline product_contract/models/sale_order.py]: Use triple double quote for docstring according PEP257.

File: product_contract/models/sale_order.py
Comment (pedrobaeza) [inline product_contract/models/sale_order.py]: Save one indentation level doing:
```
for rec in self.filtered(lambda x: x.state != 'done'):
```

File: product_contract/tests/__init__.py
Comment (pedrobaeza) [inline product_contract/tests/__init__.py]: Line at the end

File: product_contract/models/sale_order.py
Comment (pedrobaeza) [inline product_contract/models/sale_order.py]: The creation of the contract should be on `action_confirm`, not `action_done`

File: product_contract/models/sale_order.py
Comment (pedrobaeza) [reply product_contract/models/sale_order.py]: Of course, this one has no sense if you change the method where it acts.

File: product_contract/README.rst
Comment (pedrobaeza) [inline product_contract/README.rst]: Instead of contracts, you should talk about contract templates.

File: product_contract/README.rst
Comment (pedrobaeza) [inline product_contract/README.rst]: Shouldn't you add a many2one in the sales order line to keep a reference to the created contract? Even an smart-button like the one that `project` adds to the order can be good.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: product_contract/tests/__init__.py
Comment (lasley) [reply product_contract/tests/__init__.py]: Preferences => File Saving => Ensure File End With EOL Marker (assuming you're still using Komodo)

**Comment by tedsalmon** [comment]
Hey @pedrobaeza,

Thanks for the review! I've addressed all the points you've made :)

Thanks!

File: product_contract/models/sale_order.py
Comment (pedrobaeza) [inline product_contract/models/sale_order.py]: No need to put `==True`. Every Python expression is evaluated as boolean when needed

File: product_contract/models/sale_order.py
Comment (pedrobaeza) [inline product_contract/models/sale_order.py]: No need to do two loops. You can do:
```
order_lines = self.mapped('order_lines').filtered(lambda r: r.product_id.is_contract)
```

Comment (tedsalmon) [reply product_contract/models/sale_order.py]: Thank you! This is considerably more attractive to look at than my solution :)

File: product_contract/models/sale_order.py
Comment (lasley) [reply product_contract/models/sale_order.py]: Also just a note - that was nasty PEP8 too. No spaces between the `==` and the vars? What are we, back in PHP or something? 😛

File: product_contract/README.rst
Comment (lasley) [inline product_contract/README.rst]: Bad URI

File: product_contract/__manifest__.py
Comment (lasley) [inline product_contract/__manifest__.py]: Only time you need to depend on base is when you don't have any other depends. One can assume that all modules follow this principle, so at some point you will hit a base depends once you hit the bottom of the depends stack.

File: product_contract/models/product_template.py
Comment (lasley) [inline product_contract/models/product_template.py:19]: This docblock is out of spec. Take a look at [PEP-257](https://www.python.org/dev/peps/pep-0257/)

File: product_contract/models/sale_order.py
Comment (lasley) [inline product_contract/models/sale_order.py]: always return result from super. Otherwise you're breaking inheritance chain. Keep in mind that if a method doesn't return anything, it is still implicitly returning `None`. Knowing that it's returning `None` requires looking at the parent though, which is a no-no in modular design

File: product_contract/models/sale_order_line.py
Comment (lasley) [inline product_contract/models/sale_order_line.py]: No need to change, but going forward I prefer that you statically define all keys in field definitions. `comodel_name=`

**Review by lasley — APPROVED** [review]
Thanks @tedsalmon - few comments but otherwise 👍

File: product_contract/README.rst
Comment (lasley) [inline product_contract/README.rst:12]: Missing usage section with Runbot link

**Comment by lasley** [comment]
Please also make your dependencies a checklist. This will show up in the high level views as a task list that gets completed, allowing for reviewers to more easily spot PRs that are ready (and ones they should review first if they're interested in something):

image

You check the task off once the branch has been merged & you have rebased onto the production head. Now everyone is able to tell the progress without having to view the PR 🚀 

(PS I'm trying to figure out good wording for this for our Contribution docs. I'm settling with internal training first in order to figure out the verbiage after explaining a few times)

**Comment by tedsalmon** [comment]
Hey @lasley,

I've addressed all your concerns :D

**Comment by lasley** [comment]
You didn't handle my concern regarding the dependencies noted in https://github.com/OCA/contract/pull/49#issuecomment-273251203

**Comment by tedsalmon** [comment]
@lasley - I think I have now. I felt stupid trying to figure out how to use the task list *eyeroll*

**Comment by lasley** [comment]
@tedsalmon #42  has been merged, please rebase & update PR description

**Comment by tedsalmon** [comment]
@lasley,

Done! I also had to fix the tests and some functionality that broke along the way.

File: product_contract/__manifest__.py
Comment (lasley) [inline product_contract/__manifest__.py]: Remove base depend

File: product_contract/tests/__init__.py
Comment (lasley) [inline product_contract/tests/__init__.py]: NL EOF

**Review by lasley — CHANGES_REQUESTED** [review]

**Comment by tedsalmon** [comment]
Fixed -- Thanks!

**Review by lasley — APPROVED** [review]
Thanks @tedsalmon

**Comment by lasley** [comment]
@pedrobaeza would you mind checking that @tedsalmon has attended to your review comments when you get a chance?

**Review by pedrobaeza — APPROVED** [review]
Squash commits at your taste and this one is good to go

**Comment by lasley** [comment]
Squashed!

**Comment by rafaelbn** [comment]
Hi @tedsalmon , sorry for writing here but I'm unable to create a contract from sale order

```Add logic to create contracts from sale.order that contains contract products```

I've recorded a video: https://drive.google.com/file/d/1iwH27-uwvzSZhV2koIcEq8BiYrSkkRu-/view?usp=sharing

Maybe could you give me a tip?

**Comment by lasley** [comment]
@rafaelbn - looks broken to me. Seems we already have #138 for it as well. I'll take a look soon.

**Pull request merged**
